### PR TITLE
Support containers launched with user namespace

### PR DESCRIPTION
**Pull request by nbdd0121**
By default Docker/k8s run the root user inside the container as the root user of the host. This means that there's syscall filtering and path masking necessary to prevent the root of inside container from escaping. This negatively affects Nix sandboxing inside containers, because syscalls it rely on are blocked.

User namespaces also the root user inside a container to be mapped to a non-root user -- this would make all filtering and masking unnecessary (and is what rootless containers use). When we enable userns, we can then disable filtering/masking without compromising security, and then enable Nix sandbox.

### CODE REVIEW

File: src/util/namespace.rs [hunk 1]
```diff
@@ -30,9 +80,35 @@ impl MntNamespace {
...
+                    // Entering the user namespace turns out to be problematic.
+                    // The reason seems to be this line [1]:
+                    // which means `CAP_MKNOD` capability of the *init* namespace is needed.
+                    // However task's associated security context is all relative to its current
+                    // user namespace [2], so once you enter a user namespace there's no way of getting
+                    // back `CAP_MKNOD` of the init namespace anymore.
+                    // (Yes this means that even if CAP_MKNOD is granted to the container, you canot
+                    // create device nodes within it.)
+                    //
+                    // https://elixir.bootlin.com/linux/v6.11.1/source/fs/namei.c#L4073
```
Comment (pamaury) [inline src/util/namespace.rs:100]: Forgot the [1] and [2]?

File: src/util/namespace.rs [hunk 2]
```diff
@@ -30,9 +80,35 @@ impl MntNamespace {
...
+                    // namespace to be able to create files properly (otherwise EOVERFLOW
+                    // will be returned when creating file).
+                    //
+                    // Entering the user namespace turns out to be problematic.
+                    // The reason seems to be this line [1]:
+                    // which means `CAP_MKNOD` capability of the *init* namespace is needed.
+                    // However task's associated security context is all relative to its current
+                    // user namespace [2], so once you enter a user namespace there's no way of getting
+                    // back `CAP_MKNOD` of the init namespace anymore.
+                    // (Yes this means that even if CAP_MKNOD is granted to the container, you canot
```
Comment (pamaury) [inline src/util/namespace.rs:97]: typeo: cannot

File: .cargo/config.toml [hunk 3]
```diff
@@ -0,0 +1,2 @@
+[target.x86_64-unknown-linux-gnu]
+rustflags = ["-Zlinker-features=-lld"]
```
Comment (pamaury) [inline .cargo/config.toml:2]: why is this needed?

**Review by pamaury — APPROVED** [review]
This looks ok to me but I don't know enough about the details of namespaces to give some detailed feedback.

File: .cargo/config.toml (hunk 3 above)
Comment (nbdd0121) [reply .cargo/config.toml:2]: Nightly rust now default to use its bundled LLD instead of system linker, which causes issue on NixOS, this just reverts it to the stable behaviour.

Comment (nbdd0121) [reply .cargo/config.toml]: I pushed the change directly to main as it's not related to the namespace change.
